File: app/src/test/java/email/schaal/ocreader/JsonTest.java
package email.schaal.ocreader;

import android.os.Build;

import com.squareup.moshi.Moshi;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.io.IOException;

import email.schaal.ocreader.database.model.Feed;
import email.schaal.ocreader.api.json.FeedTypeAdapter;
import email.schaal.ocreader.database.model.Item;
import email.schaal.ocreader.api.json.ItemTypeAdapter;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

@RunWith(RobolectricTestRunner.class)
public class JsonTest {
    private static class ReducedItem {
        long id;
        String contentHash;
        Boolean isUnread;
        Boolean isStarred;

        @Override
        public boolean equals(Object obj) {
            if(obj instanceof ReducedItem) {
                ReducedItem other = (ReducedItem) obj;
                return id == other.id && contentHash.equals(other.contentHash) && isUnread == other.isUnread && isStarred == other.isStarred;
            }
            return super.equals(obj);
        }
    }

    @Test
    public void TestJsonWithUnexpectedNull() throws IOException {
        Moshi moshi = new Moshi.Builder().add(Feed.class, new FeedTypeAdapter()).build();
        String feedJson = "{\"id\":28,\"url\":\"http://rss.slashdot.org/Slashdot/slashdot\",\"title\":\"Slashdot\",\"faviconLink\":null,\"added\":1435334890,\"folderId\":0,\"unreadCount\":1093,\"ordering\":null,\"link\":\"http://slashdot.org/\",\"pinned\":false}";
        Feed feed = moshi.adapter(Feed.class).fromJson(feedJson);
        assertNull(feed.getFaviconLink());
    }

    @Test
    public void TestItemToJson() throws IOException {
        Moshi moshi = new Moshi.Builder().add(Item.class, new ItemTypeAdapter()).build();
        Item item = new Item(1);
        item.setContentHash("oijoijo");
        item.setUnreadChanged(true);
        item.setUnread(false);

        String itemJson = moshi.adapter(Item.class).toJson(item);
        ReducedItem reducedItem = moshi.adapter(ReducedItem.class).fromJson(itemJson);
        ReducedItem expectedReducedItem = new ReducedItem();

        expectedReducedItem.id = 1;
        expectedReducedItem.contentHash = "oijoijo";
        expectedReducedItem.isUnread = false;
        expectedReducedItem.isStarred = null;

        assertEquals(expectedReducedItem, reducedItem);
    }
}

File: app/src/test/java/email/schaal/ocreader/JsonTest.kt
/*
 * Copyright © 2020. Daniel Schaal <daniel@schaal.email>
 *
 * This file is part of ocreader.
 *
 * ocreader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ocreader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 */

package email.schaal.ocreader

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.squareup.moshi.Moshi
import email.schaal.ocreader.api.json.ItemIds
import email.schaal.ocreader.api.json.ItemMap
import email.schaal.ocreader.database.model.Item
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class JsonTest {
    @Test
    fun testItemId() {
        val items = listOf(
                Item(1),
                Item(2),
                Item(3)
        )
        val itemIds = ItemIds(items)

        val adapter = Moshi.Builder().build().adapter(ItemIds::class.java)
        val adapterJsonString = adapter.toJson(itemIds)

        Assert.assertEquals(adapter.fromJson(adapterJsonString), itemIds)
    }

    @Test
    fun testItemMap() {
        val items = listOf(
                Item(1, feedId = 11L, guidHash = "hash1"),
                Item(2, feedId = 12L, guidHash = "hash2"),
                Item(3, feedId = 13L, guidHash = "hash3")
        )

        val itemMap = ItemMap(items)

        val adapter = Moshi.Builder().build().adapter(ItemMap::class.java)
        val adapterJsonString = adapter.toJson(itemMap)

        Assert.assertEquals(adapter.fromJson(adapterJsonString), itemMap)
    }
}